### PR TITLE
See More For Tasks

### PR DESCRIPTION
**Pull request by shmbajaj**
## See More Button For Tasks 
Closes Phase-1 [367](https://github.com/Real-Dev-Squad/website-status/issues/367)

### Before The Change
- Currently, status-site  has  in-progress task  list accordion, on clicking on it shows list of tasks. 

https://user-images.githubusercontent.com/29247011/215776927-9bafb645-d78f-47d4-8223-2474d1e910f7.mp4

### After The Change
- Intially on clicking according we will be showing 3 only tasks and a see more button, and on  each click on **see more** button, we will add 5 more items to the in-progress task list. Once all tasks are rendered **see more button** will be unmounted.
- Further on closing or again opening accordion will show all tasks without **see more** button. 

https://user-images.githubusercontent.com/29247011/215778545-82330018-b7e4-430a-8e83-24429491ed6d.mp4

### Dev Tested
- Yes  [ x ] 
- No  [  ]

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -5,4 +5,6 @@ package-lock.json
 .next
 .env.local
 yarn-error.log
-coverage
\ No newline at end of file
+coverage
+.yarn
+.yarnrc.yml
```
Comment (Maheima) [inline .gitignore:10]: why is this changed?

File: src/components/Accordion/index.tsx [hunk 2]
```diff
@@ -40,9 +39,9 @@ const Accordion: VFC<Props> = ({ title, open = true, children }) => {
         <span>{isOpen ? '-' : '+'}</span>
       </div>
       <div data-testid="accordion-content" className={`${styles.content} ${isOpen ? styles.open : styles.close}`}>
-        {children}
+          {children}
```
Comment (Maheima) [inline src/components/Accordion/index.tsx:42]: why is this changed?

File: src/components/tasks/TaskList/TaskList.tsx [hunk 3]
```diff
@@ -0,0 +1,57 @@
+import beautifyTaskStatus from "@/helperFunctions/beautifyTaskStatus";
+import task from "@/interfaces/task.type";
+import { useState } from "react";
+import Card from "../card";
+import styles from "../card/card.module.scss";
```
Comment (Maheima) [inline src/components/tasks/TaskList/TaskList.tsx:5]: can we sort the imports here somewhat like in the format where
the library imports are first
then components
then CSS followed by assets

**Review by Maheima — CHANGES_REQUESTED** [review]

File: src/components/tasks/TaskList/TaskList.tsx (hunk 3 above)
Comment (shmbajaj) [reply src/components/tasks/TaskList/TaskList.tsx:5]: Imports added in order
```jsx
Library
Components
Assets
CSS
```

File: .gitignore (hunk 1 above)
Comment (shmbajaj) [reply .gitignore:10]: On `yarn install` or `yarn` it created **.yarn** directory with few script files 
and **.yarnrc.yml** which is not required. 
So, I added to gitignore.

File: src/components/Accordion/index.tsx (hunk 2 above)
Comment (shmbajaj) [reply src/components/Accordion/index.tsx:42]: I guess this is been created by `prettier`, my editor settings are different.
Hence, this make it happened.

File: .gitignore (hunk 1 above)
Comment (Maheima) [reply .gitignore:10]: @RitikJaiswal75 also faced the same issue. He can help you out with fixing it.

File: src/components/Accordion/index.tsx (hunk 2 above)
Comment (Maheima) [reply src/components/Accordion/index.tsx:42]: Can you please revert it back to how it was?

File: .gitignore (hunk 1 above)
Comment (shmbajaj) [reply .gitignore]: Reverted.

File: src/components/Accordion/index.tsx (hunk 2 above)
Comment (shmbajaj) [reply src/components/Accordion/index.tsx]: Revereted.

**Review by Maheima — APPROVED** [review]
LGTM!
Good work @shmbajaj

File: __tests__/Unit/Components/TaskList.test.tsx [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+    const seeMore = screen.getByRole("button", { name: /see more/i });
+
+    expect(seeMore).toBeInTheDocument();
+  });
+
+  it("Should render 3 tasks intially and then render more 5 tasks after click event", function () {
+    render(<TaskList tasks={tasks} hasLimit={true} />);
+
+    const _tasks = screen.queryAllByText(TASK.title);
```
Comment (heyrandhir) [inline __tests__/Unit/Components/TaskList.test.tsx:52]: any particular reason for using underscore before the variable name

File: src/components/tasks/TaskList/TaskList.tsx [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+  return (
+    <>
+      {filteredTasks.map((item: task) => (
+        <Card
+          content={item}
+          key={item.id}
+          shouldEdit={isEditable}
+          onContentChange={async (id: string, newDetails: any) =>
+            isEditable && updateCardContent?.(id, newDetails)
+          }
```
Comment (heyrandhir) [inline src/components/tasks/TaskList/TaskList.tsx:47]: NIT : Instead of writing the inline function here. We can make a separate function and use that here.

**Review by heyrandhir — APPROVED** [review]
LGTM

File: __tests__/Unit/Components/TaskList.test.tsx (hunk 4 above)
Comment (shmbajaj) [reply __tests__/Unit/Components/TaskList.test.tsx:52]: To differentiate between two set of tasks visually.

Comment (heyrandhir) [reply __tests__/Unit/Components/TaskList.test.tsx:52]: Fine usually an underscore before a variable name is used to denote that the variable is private or only meant to be used internally. If possible we should modify it so that it doesn't cause confusion.

File: src/components/tasks/TaskList/TaskList.tsx (hunk 5 above)
Comment (shmbajaj) [reply src/components/tasks/TaskList/TaskList.tsx]: [Commit Refference](https://github.com/Real-Dev-Squad/website-status/pull/404/commits/ddff0ad80fba7a54b3c644c35d5c9a3b004e083a)

File: __tests__/Unit/Components/TaskList.test.tsx (hunk 4 above)
Comment (shmbajaj) [reply __tests__/Unit/Components/TaskList.test.tsx:52]: I will change it in phase2.

File: src/components/tasks/TaskList/TaskList.tsx [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+import styles from '../card/card.module.scss'
+
+type TaksListProps = {
+  tasks: task[];
+  isEditable?: boolean;
+  hasLimit?: boolean;
+  updateCardContent?: (id: string, cardDetails: task) => void;
+};
+
+function getFilteredTasks({hasLimit, tasksLimit, tasks}: {hasLimit: boolean, tasksLimit: number, tasks: task[]}) {
```
Comment (ivinayakg) [inline src/components/tasks/TaskList/TaskList.tsx:19]: can we create a type for the parameters

File: src/components/tasks/TaskList/TaskList.tsx [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+export default function TaskList({
+  tasks,
+  updateCardContent,
+  isEditable = false,
+  hasLimit = false,
+}: TaksListProps) {
+  const initialTasksLimit = hasLimit ? INITIAL_TASKS_LIMIT : tasks.length;
+  const beautifiedTasks = beautifyTaskStatus(tasks);
+  const [tasksLimit, setTasksLimit] = useState<number>(initialTasksLimit);
+  const filteredTasks = getFilteredTasks({tasks: beautifiedTasks, hasLimit, tasksLimit});
```
Comment (ivinayakg) [inline src/components/tasks/TaskList/TaskList.tsx:33]: i dont know the full context but do check wether or not we can add memoization in this

**Review by heyrandhir — APPROVED** [review]

**Review by sahsisunny — DISMISSED** [review]
:white_check_mark: PR approved! :tada:

File: src/components/tasks/TaskList/TaskList.tsx (hunk 7 above)
Comment (shmbajaj) [reply src/components/tasks/TaskList/TaskList.tsx:33]: Using `memoization` here will be over, because state value will always be changing, 
and `see more` button will be unmounted once all tasks are rendered.

File: src/components/tasks/TaskList/TaskList.tsx (hunk 6 above)
Comment (shmbajaj) [reply src/components/tasks/TaskList/TaskList.tsx]: Done, added type `FilterTasksProps`. Check this [commit](https://github.com/Real-Dev-Squad/website-status/commit/3c011b13e671793b2880dca0a71fd45ea74ec6f2) for refference.

**Review by sahsisunny — APPROVED** [review]

**Review by yesyash — APPROVED** [review]

**Review by Maheima — APPROVED** [review]
